### PR TITLE
8193682: Infinite loop in ZipOutputStream.close()

### PR DESCRIPTION
**Pull request by raviniitw2012**
Hi all,

Please review this fix for Infinite loop in ZipOutputStream.close().
The main issue here is when ever there is an exception during close operations on GZip we are not setting the deflator to a finished state which is leading to an infinite loop when we try writing on the same GZip instance( since we use while(!def.finished()) inside the write operation).

Thanks,
Ravi

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8193682](https://bugs.openjdk.java.net/browse/JDK-8193682): Infinite loop in ZipOutputStream.close()
 * [JDK-8276305](https://bugs.openjdk.java.net/browse/JDK-8276305): Infinite loop in ZipOutputStream.close() (**CSR**) ⚠️ Issue is not open.

### Reviewers
 * [Lance Andersen](https://openjdk.java.net/census#lancea) (@LanceAndersen - **Reviewer**) ⚠️ Review applies to d911297788e316486886d7c8a8c7955beae62f3b
 * [Sean Coffey](https://openjdk.java.net/census#coffeys) (@coffeys - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5522/head:pull/5522` \
`$ git checkout pull/5522`

Update a local copy of the PR: \
`$ git checkout pull/5522` \
`$ git pull https://git.openjdk.java.net/jdk pull/5522/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5522`

View PR using the GUI difftool: \
`$ git pr show -t 5522`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5522.diff">https://git.openjdk.java.net/jdk/pull/5522.diff</a>

</details>

### CODE REVIEW

**Comment by raviniitw2012** [comment]
Reviewers: @LanceAndersen @coffeys

File: src/java.base/share/classes/java/util/zip/DeflaterOutputStream.java [hunk 1]
```diff
@@ -233,12 +233,20 @@ public void finish() throws IOException {
...
+                closed = true;
+            }
+        } catch(IOException ioe) {
+             if (usesDefaultDeflater)
                 def.end();
-            out.close();
-            closed = true;
+             out.close();
+             closed = true;
+             throw ioe;
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/DeflaterOutputStream.java:249]: Have you tried using try-finally instead?

**Comment by raviniitw2012** [comment]
Hello All .

Thanks for reviewing these changes. 
While testing ZipOutputStream:closeEntry(), I have found the same infinite loop issue is reproducible. Since closeEntry() does internally close the deflater , even though the documentation does not specify it, I think it should be fine to give the similar fix in closeEntry() of ZipOutputStream as well.  So there are total of three places where we should close deflater before throwing an exception.
1.GZipOutputStream:finish()
2.DeflaterOutputStream:close()
3.ZipOutputStream:closeEntry()

I have created a CSR explaining the changes: https://bugs.openjdk.java.net/browse/JDK-8276305

**Review by LanceAndersen — APPROVED** [review]
Hi Ravi,

The current revision looks good to me.

Best
Lance

**Comment by AlanBateman** [comment]
> Hi Ravi,
> 
> The current revision looks good to me.
> 
> Best Lance

I agree, commit d9112977 looks much better. There will be follow-up clarification needed to javadoc but they can be down with another JBS issue/PR.

**Comment by raviniitw2012** [comment]
Hi All, 

In the latest commit, I have resolved an issue in the case of ZipException in ZipOutputStream.closeEntry().
Now close the deflater in ZipOutputStream.closeEntry() only in case of an IOException and not with ZipException.

Mach5 is green : https://mach5.us.oracle.com/mdash/jobs?search=id:rreddy-jdk-20211118-1708-26358737

Thanks,
Ravi

**Review by coffeys — COMMENTED** [review]
could you move the test up one directory to java/util/zip ? I don't think it's particular to GZIP any longer. Also perhaps a rename to something like CloseDeflaterTest etc.

**Review by coffeys — APPROVED** [review]

**Pull request closed without merging**